[Infra] Add tests to presubmit

Leo-Neat opened the pull request:
Adds python test support for the presubmit script.  It will only run tests in directories that code has been changed in.

jonathanmetzman commented on infra/run_tests.py:
```diff
@@ -0,0 +1,98 @@
+import utils
+
+# A list of tests that should not be run due to run time.
+BLACK_LIST = ['bisector_test.py']
```
nit: "TEST_BLACKLIST"

jonathanmetzman replied:
Also, I like to wrap these in `set(...)` so lookup times will be constant even if list grows.

Leo-Neat replied:
Done.

jonathanmetzman commented on infra/run_tests.py:
```diff
@@ -0,0 +1,98 @@
+# A list of tests that should not be run due to run time.
+BLACK_LIST = ['bisector_test.py']
+
+
```
Sorry it's clear now what I said about only running tests on changes is ambiguous.
What I meant was that if there's any change in infra/ then run tests. This makes it unlikely that we forget to run a test but it doesn't slow down oss-fuzz project devs.
How long does it take to run every test?

Leo-Neat replied:
5 min 49 sec

jonathanmetzman replied:
In that case I think running all tests makes more sense.

jonathanmetzman commented on infra/run_tests.py:
```diff
@@ -0,0 +1,98 @@
+  failing = []
+  print('Running tests in directory: ', dir_to_run)
+  for file_name in os.listdir(dir_to_run):
+    if '_test.py' in file_name:
```
I think Python's unittest module has builtin way of discovering tests (See https://github.com/google/clusterfuzz/blob/72531be0f025005b412489e93dc9a5438b3b5919/src/local/butler/py_unittest.py#L143). We should probably use it instead.

Leo-Neat replied:
I now use this method.

jonathanmetzman commented on infra/run_tests.py:
```diff
@@ -0,0 +1,98 @@
+  failing = []
+  print('Running tests in directory: ', dir_to_run)
+  for file_name in os.listdir(dir_to_run):
+    if '_test.py' in file_name:
```
If I was to search for tests myself, I would have done `if not file_name.endswith(`_test.py`) continue` to avoid indentation.

Leo-Neat replied:
Noted.

Dor1s commented on infra/run_tests.py:
```diff
@@ -0,0 +1,98 @@
+# Copyright 2019 Google LLC
```
nit: 2020

Leo-Neat replied:
Fixed.

Dor1s commented on infra/run_tests.py:
```diff
@@ -0,0 +1,98 @@
+  return passing, failing
+
+
+def get_changed_dirs():
```
should we pass these from the presubmit script instead?

Leo-Neat replied:
Done.

Dor1s commented on infra/run_tests.py:
```diff
@@ -0,0 +1,98 @@
+    The list of directories that are affected by the change.
+  """
+  change_files, _, _ = utils.execute(
+      ['git', 'diff', '--name-only', 'origin/master'])
```
the presubmit script seems to be using `FETCH_HEAD` instead of `origin/master`. I suspect that `FETCH_HEAD` is better, is a particular PR can get behind `origin/master`, but there is a comment in the presubmit script I don't understand:

```
  # FIXME: This doesn't work if branch is behind master.
```

It would be great to figure out a single way to get the diff for testing

Leo-Neat replied:
I followed your previous comment and used the diff given by the presubmit already to be consistent. I will look into this FIXME.

Dor1s commented on infra/run_tests.py:
```diff
@@ -0,0 +1,98 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing perepo_managerissions and
+# limitations under the License.
+"""Runs the tests associated with the changes made to the repo."""
```
Can this script be used to run all the tests? I think it's better to make it that way, and let the caller to specify a subset of tests to run (e.g. the presubmit script can do that for the affected files only, assuming there is a sound way to figure out all dependencies)

Leo-Neat replied:
I moved the contents of this script to presubmit.py. All tests can be run by running `sudo python3 presubmit.py tests`

jonathanmetzman commented:
I think there's something weird with the latest patchset. I don't see the changes to presubmit.py and most of the changes I do see appear unrelated to this.

Leo-Neat commented:
> I think there's something weird with the latest patchset. I don't see the changes to presubmit.py and most of the changes I do see appear unrelated to this.

I moved most of the code to the run_test function in presubmit.py. I also changed a lot of the tests to run faster by using a pre-built do_stuff_fuzzer binary rather than building it each test.

jonathanmetzman commented on infra/repo_manager_test.py:
```diff
@@ -118,7 +118,7 @@ def test_checkout_valid_pull_request(self):
       test_repo_manager.checkout_pr('refs/pull/1757/merge')
       self.assertEqual(test_repo_manager.get_current_commit(),
-                       '2a2b11cc3d370db8f7bdf73046f3290a39615347')
+                       'e9c112b3b02debe90df1ef0395f5f6f64e3cd4db')
```
Is it just me or do we frequently need to change these hashes in unrelated pull requests? 
Are the tests breaking frequently?

Leo-Neat replied:
I submitted a PR to oss-fuzz for testing purposes. #3415

Dor1s commented on infra/cifuzz/fuzz_target_test.py:
```diff
@@ -41,26 +41,28 @@ def setUp(self):
+    with unittest.mock.patch.object(utils, 'execute',
+                                    side_effect=test_all_fail):
+      self.assertFalse(
+          self.test_target.is_reproducible('/fake/path/to/testcase'))
```
can we also assert that we had 10 calls here, e.g. `self.assertEqual(10, patch.call_count)`

Leo-Neat replied:
Added.

Dor1s commented on infra/presubmit.py:
```diff
@@ -346,9 +360,13 @@ def main():
+
   success = do_checks(changed_files)
-  return bool_to_returncode(success)
+  return run_tests() and bool_to_returncode(success)
```
it does look like there is a failure in Travis right now, which it is not complaining about: https://travis-ci.org/google/oss-fuzz/jobs/652585782

Dor1s replied:
otherwise the change looks good and I'm ready to LGTM, but need to clarify these + fix the nits I left above

Leo-Neat replied:
The tests need to be run as root because of file operations/tempfile. I updated the script to run the unit tests as root.

Dor1s commented:
@jonathanmetzman let's update https://github.com/google/oss-fuzz/blob/master/.travis.yml#L15 to have two different steps:
- presubmit (format, lint, license)
- run tests

IMO that would be slightly better to separate those, but I'm fine either way

Leo-Neat commented on infra/bisector_test.py:
```diff
@@ -29,6 +29,7 @@
 TEST_DIR_PATH = os.path.dirname(os.path.realpath(__file__))
 
 
+@unittest.skip('Test is to long to be run with presubmit.')
```
Fixed.

Leo-Neat replied:
Fixed.

Leo-Neat commented:
@Dor1s I fixed the sudo bug and added a `tests` section to the Travis matrix.

jonathanmetzman commented:
> @Dor1s I fixed the sudo bug and added a `tests` section to the Travis matrix.

Let's call it infra-tests to make meaning clearer since a lot of non-infra people use travis

Dor1s commented on infra/cifuzz/cifuzz_test.py:
```diff
@@ -28,9 +29,16 @@
+TEST_FILES_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                               'test_files')
+
+# An Example fuzzer that trips an error.
```
super nit: maybe `triggers` instead of `trips`

jonathanmetzman replied:
Another nit: "example" not "Example"

Leo-Neat replied:
Fixed to both.

Dor1s commented on infra/cifuzz/cifuzz_test.py:
```diff
@@ -114,60 +122,53 @@ def test_invalid_workspace(self):
-        self.assertNotEqual(0, len(os.listdir(build_dir)))
-        self.assertTrue(run_success)
-        self.assertTrue(bug_found)
+
```
super nit: this empty line can be removed

Leo-Neat replied:
Removed.

Dor1s commented on infra/cifuzz/cifuzz_test.py:
```diff
@@ -114,60 +122,53 @@ def test_invalid_workspace(self):
+    out_dir = os.path.join(TEST_FILES_PATH, 'out')
+    for out_file in os.listdir(out_dir):
+      out_path = os.path.join(out_dir, out_file)
+      if out_file != EXAMPLE_FUZZER:
```
```
if out_file == EXAMPLE_FUZZER:
  continue
```

and the rest of the code below can have a smaller indentation

Leo-Neat replied:
Added.

Dor1s commented on infra/cifuzz/cifuzz_test.py:
```diff
@@ -114,60 +122,53 @@ def test_invalid_workspace(self):
   """Test build_fuzzers function in the cifuzz module."""
 
+  def tearDown(self):
+    """Remove any existing crashes and test files."""
```
why do we need to delete all of those? maybe we can be more concrete and delete only when needed, instead of deleting everything that is not fuzz target binary?

Leo-Neat replied:
I don't want to corrupt the sanity of the other tests.

Dor1s commented on infra/presubmit.py:
```diff
@@ -346,8 +360,12 @@ def main():
+  if args.command == 'test':
+    return run_tests()
+
   # Otherwise, do all of them.
```
nit: can you update this comment to say "do all the checks (but no tests)"?

Leo-Neat replied:
Added.
